### PR TITLE
SNOW-996732: improve clean up logic of connection

### PR DESCRIPTION
**Pull request by sfc-gh-aling**
Please answer these questions before submitting your pull requests. Thanks!

1. What GitHub issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

  the `__del__` method is unstable and gets called at random timing.
in one of the recent report, we found the `__del__` is causing circular import issue.

as a short term fix, we fix by moving away from `__del__` method to using `atexit` to clean up connection at intepreter shutdown time.

long term fix plan here: SNOW-1218851

2. Fill out the following pre-review checklist:

   - [ ] I am adding a new automated test(s) to verify correctness of my new code
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am modifying authorization mechanisms
   - [ ] I am adding new credentials
   - [ ] I am modifying OCSP code
   - [ ] I am adding a new dependency

3. Please describe how your code solves the related issue.

   Please write a short description of how your code change solves the related issue.

### CODE REVIEW

File: src/snowflake/connector/connection.py [hunk 1]
```diff
@@ -1749,6 +1747,12 @@ def _cache_query_status(self, sf_qid: str, status_ret: QueryStatus) -> None:
             )  # Prevent KeyError when multiple threads try to remove the same query id
             self._done_async_sfqids[sf_qid] = None
 
+    def _close_at_exit(self):
+        try:
```
Comment (sfc-gh-bwarsaw) [inline src/snowflake/connector/connection.py:1751]: Maybe use [contextlib.suppress()](https://docs.python.org/3/library/contextlib.html#contextlib.suppress)?

**Comment by sfc-gh-bwarsaw** [comment]
As mentioned in Slack: One thing to keep in mind is that `atexit` handlers don't get run if `os._exit()` is called or the process crashes.

File: src/snowflake/connector/connection.py (hunk 1 above)
Comment (sfc-gh-aling) [reply src/snowflake/connector/connection.py]: this is elegant

**Comment by sfc-gh-aling** [comment]
thanks @sfc-gh-bwarsaw , this should not be a concern to our connector as we didn't perform critical job during the closing stage.

File: src/snowflake/connector/connection.py [hunk 2]
```diff
@@ -439,12 +441,7 @@ def __init__(
 
         # get the imported modules from sys.modules
         self._log_telemetry_imported_packages()
-
-    def __del__(self) -> None:  # pragma: no cover
-        try:
-            self.close(retry=False)
-        except Exception:
-            pass
+        atexit.register(self._close_at_exit)
```
Comment (sfc-gh-jrose) [inline src/snowflake/connector/connection.py:444]: Should there be a comment with the long term ticket number so it's clear this is a temporary solution?

**Review by sfc-gh-jrose — APPROVED** [review]

File: src/snowflake/connector/connection.py (hunk 2 above)
Comment (sfc-gh-aling) [inline src/snowflake/connector/connection.py:444]:
```suggestion
        # check SNOW-1218851 for long term improvement plan
        atexit.register(self._close_at_exit)
```

Comment (sfc-gh-aling) [reply src/snowflake/connector/connection.py:445]: updated

File: test/integ/test_connection.py [hunk 3]
```diff
@@ -1342,3 +1342,20 @@ def test_ocsp_mode_insecure(conn_cnx, is_public_test, caplog):
...
+        password=db_parameters["password"],
+        host=db_parameters["host"],
+        port=db_parameters["port"],
+        account=db_parameters["account"],
+        database=db_parameters["database"],
+        protocol=db_parameters["protocol"],
+        timezone="UTC",
+    )
+    cnx._close_at_exit()
+    assert cnx.is_closed()
```
Comment (sfc-gh-mkeller) [inline test/integ/test_connection.py:1361]: I don't like when individual tests handle `db_parameters`, use `conn_cnx` instead! There's a smaller chance to leak the dr aprameters that way.

File: test/integ/test_connection.py [hunk 4]
```diff
@@ -1342,3 +1342,20 @@ def test_ocsp_mode_insecure(conn_cnx, is_public_test, caplog):
             assert "snowflake.connector.ocsp_snowflake" in caplog.text
         else:
             assert "snowflake.connector.ocsp_snowflake" not in caplog.text
+
+
+@pytest.mark.skipolddriver
+def test_connection_atexit_close(db_parameters, capsys):
```
Comment (sfc-gh-mkeller) [inline test/integ/test_connection.py:1348]: `capsys` is unused
```suggestion
def test_connection_atexit_close(db_parameters):
```

File: test/integ/test_connection.py (hunk 3 above)
Comment (sfc-gh-aling) [reply test/integ/test_connection.py]: updated, thanks for the feedback!

**Pull request merged**